### PR TITLE
enable deduplicating searched candidates and prune invalid states

### PR DESCRIPTION
**Pull request by CtfGo**
1. enable to deduplicate searched candidates: design two ir::IRVisitor derived class `IrNodesStructuralHash` and `ir::IrEqualVisitor` as hash functor and equal functor respectively  for `std::unordered_set`
2. add GPU memory usage limit check to prune invalid candidates

### CODE REVIEW

**Review by BiynXu — DISMISSED** [review]
LGTM

File: cinn/auto_schedule/search_space/search_state.cc [hunk 1]
```diff
@@ -61,5 +63,88 @@ bool operator<(const SearchState& left, const SearchState& right) {
   return left->predicted_cost < right->predicted_cost;
 }
 
+// Visit every node by expanding all of their fields in dfs order
```
Comment (zhhsplendid) [inline cinn/auto_schedule/search_space/search_state.cc:66]: Do you aim to include the `operation` of `ir::Tensor`?

**Review by zhhsplendid — DISMISSED** [review]
LGTM

File: cinn/auto_schedule/search_space/search_state.cc (hunk 1 above)
Comment (CtfGo) [reply cinn/auto_schedule/search_space/search_state.cc:66]: The derived class `IrNodesStructuralHash` doesn't and it will be used finally.

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**